src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs

```diff
@@ -10,53 +11,72 @@ namespace Shouldly
     public static partial class ShouldBeTestExtensions
     {
         [ContractAnnotation("actual:null,expected:notnull => halt;actual:notnull,expected:null => halt")]
-        public static void ShouldBe<T>(this T actual, T expected)
+        public static void ShouldBe<T>(
+            [AllowNull, NotNullIfNotNull("expected")] this T actual,
+            [AllowNull, NotNullIfNotNull("actual")] T expected)
         {
             ShouldBe(actual, expected, () => null);
         }
         [ContractAnnotation("actual:null,expected:notnull => halt;actual:notnull,expected:null => halt")]
-        public static void  ShouldBe<T>(this T actual, T expected, string customMessage)
+        public static void  ShouldBe<T>(
+            [AllowNull, NotNullIfNotNull("expected")] this T actual,
+            [AllowNull, NotNullIfNotNull("actual")] T expected
+            , string? customMessage)
         {
             ShouldBe(actual, expected, () => customMessage);
         }
         [ContractAnnotation("actual:null,expected:notnull => halt;actual:notnull,expected:null => halt")]
-        public static void ShouldBe<T>(this T actual, T expected, [InstantHandle] Func<string> customMessage)
+        public static void ShouldBe<T>(
+            [AllowNull, NotNullIfNotNull("expected")] this T actual,
+            [AllowNull, NotNullIfNotNull("actual")] T expected,
+            [InstantHandle] Func<string?>? customMessage)
         {
-            if (ShouldlyConfiguration.CompareAsObjectTypes.Contains(typeof(T).FullName) || typeof(T) == typeof(string))
+            if (ShouldlyConfiguration.CompareAsObjectTypes.Contains(typeof(T).FullName!) || typeof(T) == typeof(string))
                 actual.AssertAwesomely(v => Is.Equal(v, expected, new ObjectEqualityComparer<T>()), actual, expected, customMessage);
             else
                 actual.AssertAwesomely(v => Is.Equal(v, expected), actual, expected, customMessage);
         }
 
         [ContractAnnotation("actual:null,expected:null => halt")]
-        public static void ShouldNotBe<T>(this T actual, T expected)
+        public static void ShouldNotBe<T>([AllowNull] this T actual, [AllowNull] T expected)
```

Comment:
`[AllowNull]` is needed on each so that generic inference doesn't get upset when one expression is nullable and the other isn't.